thin-replica-server/test/thin_replica_server_test.cpp

```diff
@@ -954,6 +970,84 @@ TEST(thin_replica_server_test, getClientIdFromClientCert) {
+  auto total_event_groups = kLastEventGroupId + 5;
+  auto total_updates = total_legacy_blocks + total_event_groups;
+
+  // Setup ThinReplicaServer
+  bool is_insecure_trs = true;
+  std::string tls_trs_cert_path;
+  std::unordered_set<std::string> client_id_set;
+  uint16_t update_metrics_aggregator_thresh = 100;
+
+  auto trs_config = std::make_unique<concord::thin_replica::ThinReplicaServerConfig>(
+      is_insecure_trs, tls_trs_cert_path, &storage, buffer, client_id_set, update_metrics_aggregator_thresh);
+  concord::thin_replica::ThinReplicaImpl replica(std::move(trs_config), std::make_shared<concordMetrics::Aggregator>());
+  TestServerContext context;
+
+  // ThinReplica legacy events request
+  SubscriptionRequest request;
+  request.mutable_events()->set_block_id(1);
+
+  auto status =
+      replica.SubscribeToUpdates<TestServerContext, TestServerWriter<Hash>, Hash>(&context, &request, &stream);
+  EXPECT_EQ(status.error_code(), grpc::StatusCode::OK);
+  EXPECT_EQ(state_machine.numUpdatesReceived(), total_updates);
+  EXPECT_EQ(state_machine.numLegacyBlocksReceived(), total_legacy_blocks);
+  EXPECT_EQ(state_machine.numEventGroupsReceived(), total_event_groups);
+}
+
+// Subscribing to legacy events with storage that contains event groups only returns event groups
+TEST(thin_replica_server_test, SubscribeToUpdatesLegacyRequestEventGroups) {
+  // Storage contains event groups only
+  FakeStorage storage(generateEventGroupMap(1, kLastEventGroupId));
+  EXPECT_EQ(storage.getLastEventGroupId(), kLastEventGroupId);
+  EXPECT_EQ(storage.getFirstEventGroupBlockId(), 1);
+
+  // Live updates can contain event groups only
+  auto live_update_event_groups = generateEventGroupMap(kLastEventGroupId + 1, kLastEventGroupId + 10);
+  TestStateMachine<Hash> state_machine{storage, live_update_event_groups, 1};
+  TestSubBufferList<Hash> buffer{state_machine};
+  TestServerWriter<Hash> stream{state_machine};
+
+  auto total_event_groups = kLastEventGroupId + 10;
+
+  // Setup ThinReplicaServer
+  bool is_insecure_trs = true;
+  std::string tls_trs_cert_path;
+  std::unordered_set<std::string> client_id_set;
+  uint16_t update_metrics_aggregator_thresh = 100;
+
+  auto trs_config = std::make_unique<concord::thin_replica::ThinReplicaServerConfig>(
+      is_insecure_trs, tls_trs_cert_path, &storage, buffer, client_id_set, update_metrics_aggregator_thresh);
+  concord::thin_replica::ThinReplicaImpl replica_(std::move(trs_config),
+                                                  std::make_shared<concordMetrics::Aggregator>());
+  TestServerContext context;
+
+  // ThinReplica legacy events request
+  SubscriptionRequest request;
+  request.mutable_events()->set_block_id(3);
+
+  auto status =
+      replica_.SubscribeToUpdates<TestServerContext, TestServerWriter<Hash>, Hash>(&context, &request, &stream);
+  EXPECT_EQ(status.error_code(), grpc::StatusCode::OK);
+  EXPECT_EQ(state_machine.numEventGroupsReceived(), total_event_groups);
+}
```

Comment:
Nit: It might be useful to check if `numLegacyBlocksReceived()` returns 0, but that can be added separately.

Reply:
Good point! This change doesn't have immediate urgency so I will fix it now.